## Downloading Dataset

In [2]:
!wget https://zenodo.org/record/2787612/files/SICK.zip?download=1

--2023-08-02 13:05:14--  https://zenodo.org/record/2787612/files/SICK.zip?download=1
Resolving zenodo.org (zenodo.org)... 188.185.124.72
Connecting to zenodo.org (zenodo.org)|188.185.124.72|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 217584 (212K) [application/octet-stream]
Saving to: ‘SICK.zip?download=1’

SICK.zip?download=1 100%[===================>] 212.48K   578KB/s    in 0.4s    

2023-08-02 13:05:16 (578 KB/s) - ‘SICK.zip?download=1’ saved [217584/217584]



In [3]:
!unzip SICK.zip?download=1

Archive:  SICK.zip?download=1
  inflating: readme.txt              
  inflating: SICK.txt                


## Importing Libraries

In [4]:
import os
import csv
import math
import numpy as np
import pandas as pd
import seaborn as sns
from datetime import datetime
import matplotlib.pyplot as plt
from torch.utils.data import DataLoader
from sentence_transformers.evaluation import EmbeddingSimilarityEvaluator
from sentence_transformers import SentenceTransformer, LoggingHandler, losses, util, InputExample

In [6]:
df=pd.read_csv('/content/SICK.txt', delimiter='\t', encoding='utf8')
df.head()

pair_ID                                         sentence_A  \
0        1  A group of kids is playing in a yard and an ol...   
1        2  A group of children is playing in the house an...   
2        3  The young boys are playing outdoors and the ma...   
3        4  The young boys are playing outdoors and the ma...   
4        5  The kids are playing outdoors near a man with ...   

                                          sentence_B entailment_label  \
0  A group of boys in a yard is playing and a man...          NEUTRAL   
1  A group of kids is playing in a yard and an ol...          NEUTRAL   
2  The kids are playing outdoors near a man with ...       ENTAILMENT   
3  There is no boy playing outdoors and there is ...    CONTRADICTION   
4  A group of kids is playing in a yard and an ol...          NEUTRAL   

   relatedness_score    entailment_AB    entailment_BA  \
0                4.5      A_neutral_B      B_neutral_A   
1                3.2  A_contradicts_B      B_neutral_A   
2                4.7      A_entails_B      B_entails_A   
3                3.6  A_contradicts_B  B_contradicts_A   
4                3.4      A_neutral_B      B_neutral_A   

                                 sentence_A_original  \
0  A group of children playing in a yard, a man i...   
1  A group of children playing in a yard, a man i...   
2  The children are playing outdoors, while a man...   
3  The children are playing outdoors, while a man...   
4  A group of children playing in a yard, a man i...   

                                 sentence_B_original sentence_A_dataset  \
0  A group of children playing in a yard, a man i...             FLICKR   
1  A group of children playing in a yard, a man i...             FLICKR   
2  The children are playing outdoors, while a man...             FLICKR   
3  The children are playing outdoors, while a man...             FLICKR   
4  The children are playing outdoors, while a man...             FLICKR   

  sentence_B_dataset SemEval_set  
0             FLICKR       TRAIN  
1             FLICKR       TRAIN  
2             FLICKR       TRAIN  
3             FLICKR       TRIAL  
4             FLICKR       TRAIN

In [7]:
df.isnull().sum()

pair_ID                0
sentence_A             0
sentence_B             0
entailment_label       0
relatedness_score      0
entailment_AB          0
entailment_BA          0
sentence_A_original    0
sentence_B_original    0
sentence_A_dataset     0
sentence_B_dataset     0
SemEval_set            0
dtype: int64

In [8]:
df.dtypes

pair_ID                  int64
sentence_A              object
sentence_B              object
entailment_label        object
relatedness_score      float64
entailment_AB           object
entailment_BA           object
sentence_A_original     object
sentence_B_original     object
sentence_A_dataset      object
sentence_B_dataset      object
SemEval_set             object
dtype: object

In [9]:
df.SemEval_set.unique()

array(['TRAIN', 'TRIAL', 'TEST'], dtype=object)

## Preparing Input Sentences for Training

In [10]:
dataset_path = '/content/SICK.txt'

# Using pretrained model
model_name = 'nli-distilroberta-base-v2'
train_batch_size = 16
model_save_path = 'output/training-'+model_name+'-'+datetime.now().strftime("%Y-%m-%d_%H-%M-%S")

# Loading a pre-trained sentence transformer model
model = SentenceTransformer(model_name)

# Preparing data for training objectives
train_samples = []
dev_samples = []
test_samples = []
with open(dataset_path, 'rt', encoding='utf8') as fIn:
    reader = csv.DictReader(fIn, delimiter='\t', quoting=csv.QUOTE_NONE)
    for row in reader:
        score = float(row['relatedness_score'])/ 5.0  # Normalize score to range 0 ... 1
        inp_example = InputExample(texts=[row['sentence_A'], row['sentence_B']], label=score)
        if row['SemEval_set'] == 'TRIAL':
            dev_samples.append(inp_example)
        elif row['SemEval_set'] == 'TEST':
            test_samples.append(inp_example)
        else:
            train_samples.append(inp_example)


## Performance of base model

In [13]:
test_evaluator = EmbeddingSimilarityEvaluator.from_input_examples(test_samples, name='sick-test')
print('Base model performance:', test_evaluator(model))

Base model performance: 0.7877062927879224


## Fine-tuning our model for better embeddings

In [15]:
train_dataloader = DataLoader(train_samples, shuffle=True, batch_size=train_batch_size)
train_loss = losses.CosineSimilarityLoss(model=model)


# Evaluate a model based on the similarity of the embeddings by calculating the Spearman and Pearson rank correlation
#in comparison to the gold standard labels.
evaluator = EmbeddingSimilarityEvaluator.from_input_examples(dev_samples, name='sick-dev')


num_epochs=20
# Training the fine tuned model
model.fit(train_objectives=[(train_dataloader, train_loss)],
          evaluator=evaluator,
          epochs=num_epochs,
          evaluation_steps=1000,
          output_path=model_save_path)


new_model = SentenceTransformer(model_save_path)

Epoch:   0%|          | 0/20 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

Iteration:   0%|          | 0/278 [00:00<?, ?it/s]

## Performance of new fine-tuned model

In [17]:
test_evaluator = EmbeddingSimilarityEvaluator.from_input_examples(test_samples, name='sick-test')
print('New fined-tuned model performance:', test_evaluator(new_model, output_path=model_save_path))

New fined-tuned model performance: 0.8474205455002625


* We can see that after fine-tuning our model's performace increased from base 0.787 to 0.847

### Few examples of similarity between sentences using our model

In [18]:
def cosine_sim(sent1, sent2):
  emb=new_model.encode([sent1,sent2])
  sen1_emb=emb[0]
  sent2_emb=emb[1]
  cos_sim=np.dot(emb[0],emb[1])/(np.linalg.norm(emb[0])*np.linalg.norm(emb[1]))
  return cos_sim

In [19]:
# Few cosine similarity examples
q=cosine_sim('How do I read my YouTube comments?',	'How do I see my YouTube comments?')
if q>0.90:
  print(f'The sentences are similar with cosine score',q)
else:
  print('They sentences have difference, the cosine score is',q)

The sentences are similar with cosine score 0.964395


In [20]:
q=cosine_sim('Is God existence itself?	', 'Does God exist?')
if q>0.90:
  print(f'The sentences are similar with cosine score',q)
else:
  print('They sentences have difference, the cosine score is',q)

The sentences are similar with cosine score 0.9530106


In [21]:
q=cosine_sim('The boys played very well in match.', 'The boys did not played very well in match.')
if q>0.90:
  print(f'The sentences are similar with cosine score',q)
else:
  print('They sentences have difference, the cosine score is',q)

They sentences have difference, the cosine score is 0.74921983


## Comparing our model with most downloaded model in Hugging-face library for sentence-similarity

In [22]:
# Comparing our fine-tuned model with the most popular sentence-similarity model from hugging face library
from sentence_transformers import SentenceTransformer
sentences = ['How do I read my YouTube comments?',	'How do I see my YouTube comments?']

popular_model = SentenceTransformer('sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2')

In [24]:
test_evaluator = EmbeddingSimilarityEvaluator.from_input_examples(test_samples, name='sts-test')
print('Most popular model simiarity performance:',test_evaluator(popular_model))

Most popular model simiarity performance: 0.7924763887870493


* We can that our fine-tuned model exceeded the performace of the most popular model in hugging-face library `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2` by 6.94%

In [36]:
# Similarity evaluation results for our fine-tuned model
test_result=pd.read_csv('/content/output/training-nli-distilroberta-base-v2-2023-08-02_13-19-28/similarity_evaluation_sick-test_results.csv')
test_result

epoch  steps  cosine_pearson  cosine_spearman  euclidean_pearson  \
0     -1     -1        0.893237         0.847421           0.869195   

   euclidean_spearman  manhattan_pearson  manhattan_spearman  dot_pearson  \
0            0.839976           0.869092            0.839242     0.834107   

   dot_spearman  
0      0.767781

No charts were generated by quickchart
